ghc-lib-gen/src/Ghclibgen.hs

```diff
@@ -854,8 +892,14 @@ generateGhcLibCabal ghcFlavor = do
         , "homepage: https://github.com/digital-asset/ghc-lib"
         , "bug-reports: https://github.com/digital-asset/ghc-lib/issues"
         , "data-dir: " ++ dataDir
-        , "data-files:"] ++ indent dataFiles ++
-        [ "extra-source-files:"] ++ indent (ghcLibExtraFiles ghcFlavor \\ [hadrianGeneratedRoot ghcFlavor </> "ghcversion.h"]) ++
+        , "data-files:"] ++ indent (dataFiles ghcFlavor) ++
+        [ "extra-source-files:"] ++
+          indent (ghcLibExtraFiles ghcFlavor
+                    [ (hadrianGeneratedRoot ghcFlavor </> "ghcversion.h", Nothing)
+                    , (hadrianGeneratedRoot ghcFlavor </> "DerivedConstants.h", Just $ hadrianGeneratedRoot ghcFlavor </> "GhclibDerivedConstants.h")
+                    , ("compiler" </> "HsVersions.h", Just $ "compiler" </> "GhclibHsVersions.h")
```

Comment:
Seems identical to the list below, let’s factor it out.

Reply:
ok. to address these two points,  `ghcLibParserExtraFiles` and `ghcLibExtraFiles` will no longer have substituion logic. apply substitutions in new function `performExtraFilesSubsitutions` and only when there are substitutions to apply.